include/mbgl/util/image.hpp

```diff
@@ -58,6 +71,50 @@ class Image : private util::noncopyable {
     size_t stride() const { return channels * size.width; }
     size_t bytes() const { return stride() * size.height; }
 
+    void fill(uint8_t value) {
+        std::fill(data.get(), data.get() + bytes(), value);
+    }
+
+    // Copy image data within `rect` from `src` to the rectangle of the same size at `pt`
+    // in `dst`. If the specified bounds exceed the bounds of the source or destination,
+    // throw `std::out_of_range`. Must not be used to move data within a single Image.
+    static void copy(const Image& srcImg, Image& dstImg, const Point<uint32_t>& srcPt, const Point<uint32_t>& dstPt, const Size& size) {
+        if (!srcImg.valid()) {
+            throw std::invalid_argument("invalid source for image copy");
+        }
+
+        if (!dstImg.valid()) {
+            throw std::invalid_argument("invalid destination for image copy");
+        }
+
+        if (size.width > srcImg.size.width ||
+            size.height > srcImg.size.height ||
+            srcPt.x > srcImg.size.width - size.width ||
+            srcPt.y > srcImg.size.height - size.height) {
+            throw std::out_of_range("out of range source coordinates for image copy");
+        }
+
+        if (size.width > dstImg.size.width ||
+            size.height > dstImg.size.height ||
+            dstPt.x > dstImg.size.width - size.width ||
+            dstPt.y > dstImg.size.height - size.height) {
+            throw std::out_of_range("out of range destination coordinates for image copy");
+        }
```

Comment:
Instead of failing hard in the out of range case, we could also clip to the common intersection, which is what Canvas does.